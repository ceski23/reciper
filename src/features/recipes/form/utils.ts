import { group } from 'radash'
import { type RecipeFormValues } from 'features/recipes/form/scheme'
import { type Recipe } from 'features/recipes/types'

export const mapFormValuesToRecipe = (values: RecipeFormValues) => ({
	...values,
	ingredients: values.ingredients.flatMap(({ items, group }) =>
		items.map(item => ({
			...item,
			group,
		}))
	),
	instructions: values.instructions.flatMap(({ items, group }) =>
		items.map(item => ({
			...item,
			group,
		}))
	),
	tags: values.tags.map(tag => tag.text),
} satisfies Omit<Recipe, 'id' | 'addedDate'>)

export const mapRecipeToFormValues = (recipe: Recipe) => ({
	...recipe,
	ingredients: Object.entries(group(recipe.ingredients, item => item.group ?? '')).map(([group, items]) => ({
		group: group || undefined,
		items: items?.map(({
			// eslint-disable-next-line @typescript-eslint/no-unused-vars
			group,
			...item
		}) => item) ?? [],
	})),
	instructions: Object.entries(group(recipe.instructions, item => item.group ?? '')).map(([group, items]) => ({
		group: group || undefined,
		items: items?.map(({
			// eslint-disable-next-line @typescript-eslint/no-unused-vars
			group,
			...item
		}) => item) ?? [],
	})),
	tags: recipe.tags.map(tag => ({ text: tag })),
} satisfies RecipeFormValues)
