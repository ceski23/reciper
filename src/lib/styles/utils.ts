import { type Breakpoint, breakpoints } from '@styles/theme'

export const transparentize = (color: string, opacity: number) => `color-mix(in srgb, ${color}, transparent ${100 - (opacity * 100)}%)`
export const blendWithColor = (color: string, color2: string, opacity: number) => `color-mix(in srgb, ${color}, ${color2} ${opacity * 100}%)`
export const mergeClassNames = (classNames: Array<string | undefined>) =>
	classNames.filter(className => className !== undefined && className.length > 0).join(' ')

export const mq = {
	upTo: (breakpoint: Breakpoint) => `(${breakpoints[breakpoint]}px > width)`,
	atLeast: (breakpoint: Breakpoint) => `(${breakpoints[breakpoint]}px <= width)`,
	between: (breakpointA: Breakpoint, breakpointB: Breakpoint) => `(${breakpoints[breakpointA]}px <= width < ${breakpoints[breakpointB]}px)`,
	only: (breakpoint: Breakpoint) => {
		const entries = Object.entries(breakpoints) as Array<[keyof typeof breakpoints, number]>
		const index = entries.findIndex(([key]) => key === breakpoint)
		const nextBreakpoint = entries.at(index + 1)?.[1]

		return nextBreakpoint !== undefined
			? `(${breakpoints[breakpoint]}px <= width < ${nextBreakpoint}px)`
			: `(${breakpoints[breakpoint]}px <= width)`
	},
	darkMode: '(prefers-color-scheme: dark)',
}
