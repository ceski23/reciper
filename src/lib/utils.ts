export const isDefined = <T>(item: T): item is NonNullable<T> => item !== undefined && item !== null
