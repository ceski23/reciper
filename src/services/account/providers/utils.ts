export const randomBytes = (size: number) => {
  return crypto.getRandomValues(new Uint8Array(size))
}

export const base64url = (bytes: any) => {
  return btoa(String.fromCharCode(...bytes))
    .replace(/=/g, '')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
}

export const generateCodeChallenge = async (code_verifier: string) => {
  const codeVerifierBytes = new TextEncoder().encode(code_verifier)
  const hashBuffer = await crypto.subtle.digest('SHA-256', codeVerifierBytes)
  return base64url(new Uint8Array(hashBuffer))
}