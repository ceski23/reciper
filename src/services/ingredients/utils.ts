export const kgToG = 1000;
export const gToKg = 1 / 1000;

export const lToMl = 1000;
export const mlToL = 1 / 1000;
