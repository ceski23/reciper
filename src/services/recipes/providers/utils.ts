import Vibrant from 'node-vibrant';

export const colorExtractor = async (imageUrl: string) => {
  const palette = await Vibrant.from((import.meta.env.VITE_CORS_PROXY as string ?? '') + encodeURIComponent(imageUrl)).getPalette();
  return palette;
};
