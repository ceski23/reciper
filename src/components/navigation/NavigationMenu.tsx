import styled from '@emotion/styled';
import { FC } from 'react';

import { media } from 'utils/styles/mediaQueries';
import { color } from 'utils/styles/theme';

const NavigationContainer = styled.nav`
  display: flex;
  flex-direction: row;
  background-color: ${color('background')};
  width: 100%;
  z-index: 10;

  ${media.down('medium')} {
    box-shadow: 0px 0px 20px ${color('shadow')};
    height: 70px;
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
  }
`;

const MenuItems = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  padding: 0;
  margin: 0;
  width: 100%;
  height: 100%;
  gap: 10px;
  padding: 10px;

  ${media.down('medium')} {
    flex-direction: row;
    padding: 0;
  }
`;

export const NavigationMenu: FC = ({ children }) => (
  <NavigationContainer>
    <MenuItems>
      {children}
    </MenuItems>
  </NavigationContainer>
);
