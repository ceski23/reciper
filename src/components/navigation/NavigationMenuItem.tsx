/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react/macro';
import styled from '@emotion/styled/macro';
import { transparentize } from 'polished';
import { FC } from 'react';
import { NavLink, NavLinkProps } from 'react-router-dom';

import { media } from 'utils/styles/mediaQueries';

const Item = styled(NavLink)`
  display: flex;
  flex-direction: row;
  align-items: center;
  padding: 15px;
  position: relative;
  text-decoration: none;
  color: unset;
  font-weight: 600;
  gap: 10px;
  border-radius: 10px;

  &.active {
    color: ${(props) => props.theme.colors.primary};
    background-color: ${(props) => transparentize(0.9, props.theme.colors.primary)};
  }

  ${media.down('medium')} {
    &.active {
      background-color: transparent;

      &::after {
        content: '';
        display: block;
        position: absolute;
        bottom: 0;
        width: 80%;
        height: 4px;
        left: 0;
        right: 0;
        margin: auto;
        border-radius: 15px 15px 0 0;
        background-color: currentColor;
      }
    }
  }
`;

const Text = styled.span`
  display: none;

  ${media.up('small')} {
    display: block;
  }
`;

const iconStyles = css`
  width: 25px;
  fill: currentColor;
`;

interface Props {
  title: string;
  to: NavLinkProps['to'];
  icon?: React.FunctionComponent<React.SVGProps<SVGSVGElement>>;
}

export const NavigationMenuItem: FC<Props> = ({ title, to, icon: Icon }) => (
  <Item to={to}>
    {Icon && <Icon css={iconStyles} />}
    <Text>{title}</Text>
  </Item>
);
