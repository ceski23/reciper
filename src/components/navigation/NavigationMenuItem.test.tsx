import { renderWithProviders } from 'test/utils';

import { NavigationMenuItem } from 'components/navigation/NavigationMenuItem';

describe('NavigationMenuItem', () => {
  it('should render correctly', () => {
    const { asFragment } = renderWithProviders(
      <NavigationMenuItem title="Test" to="/test" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
