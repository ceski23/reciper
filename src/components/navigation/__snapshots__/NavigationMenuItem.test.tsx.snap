// Vitest Snapshot v1

exports[`NavigationMenuItem > should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 15px;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: unset;
  font-weight: 600;
  gap: 10px;
  border-radius: 10px;
}

.emotion-0.active {
  color: #0af;
  background-color: rgba(0,170,255,0.1);
}

@media (max-width: 768px) {
  .emotion-0.active {
    background-color: transparent;
  }

  .emotion-0.active::after {
    content: '';
    display: block;
    position: absolute;
    bottom: 0;
    width: 80%;
    height: 4px;
    left: 0;
    right: 0;
    margin: auto;
    border-radius: 15px 15px 0 0;
    background-color: currentColor;
  }
}

.emotion-2 {
  display: none;
}

@media (min-width: 576px) {
  .emotion-2 {
    display: block;
  }
}

<a
    class="emotion-0 emotion-1"
    href="/test"
  >
    <span
      class="emotion-2 emotion-3"
    >
      Test
    </span>
  </a>
</DocumentFragment>
`;
